Allow setting Jinja environment arguments from settings

rjames86 opened the pull request:
Allow the ability to set custom Jinja environment arguments from within the settings. One consideration is that this setting could also be used in favor of `"JINJA_EXTENSIONS"` by just setting the key in the dictionary.

One potential downside is that someone could override the arguments already being set [here](https://github.com/getpelican/pelican/blob/master/pelican/generators.py#L64). We could whitelist/blacklist arguments that need to stay as the defaults.

justinmayer commented:
Thanks for the contribution, Ryan. It looks like you may have forgotten to run `flake8` on the files you changed, which ensures that PyCodeStyle guidelines are followed. This is something our test suite checks for, as you can see by the failing check in this pull request.

@avaris / @ingwinlu / @leotrs: Any thoughts or comments?

avaris commented:
If we are going this way, we might as well deprecate `JINJA_EXTENSIONS` and set the predetermined values (`trim_blocks`, `lstrip_blocks`) as defaults in this setting.

rjames86 commented:
@avaris Sounds good. I figured that would be the best way forward, but wanted to check to be sure. I'll get a change out for that

rjames86 commented:
One question: I've removed the `JINJA_EXTENSIONS` setting and updated the docs. Are there any other places that I need to update for deprecated settings variables?

avaris commented:
I think the wording on that warning might be confusing. To me, it sounds like only the setting name is changed, but it's a completely new setting with a different context. Maybe something like:

```
JINJA_EXTENSIONS setting has been deprecated, moving it to JINJA_ENVIRONMENT setting.
```

justinmayer commented:
@avaris: Any comments on Ryan's latest changes? Anything else here that you think needs to be addressed?

avaris commented on pelican/settings.py:
```diff
@@ -212,6 +222,18 @@ def get_settings_from_file(path, default_settings=DEFAULT_CONFIG):
+def get_jinja_environment(settings):
+    """Sets the environment for Jinja"""
+
+    jinja_env = settings.get('JINJA_ENVIRONMENT', {})
```
Slight issue. `.get` here doesn't do anything if the default is returned. You'll get an empty dict which is then populated but nothing will happen to the `settings`. If anything, `.setdefault` should be used instead.

Since any settings will use `DEFAULT_CONFIG` as base, there should be a `'JINJA_ENVIRONMENT'` key regardless, but still :).

justinmayer commented:
@rjames86: Could you take another pass based on the feedback from @avaris?

rjames86 commented:
@justinmayer yep will do. was planning on doing it this weekend

avaris commented on pelican/settings.py:
```diff
@@ -212,6 +222,21 @@ def get_settings_from_file(path, default_settings=DEFAULT_CONFIG):
+def get_jinja_environment(settings):
+    """Sets the environment for Jinja"""
+
+    jinja_env = settings['JINJA_ENVIRONMENT']
```
This is unnecessary and the cause of test failures. `.setdefault` will return the value if it exists, else sets the value and returns it.

rjames86 replied:
Whoops. Sorry, that's my fault. That wasn't supposed to be there

avaris commented:
This is good to go once you [squash](https://github.com/getpelican/pelican/wiki/Git-Tips#squashing-commits) it :).

rjames86 commented:
@avaris just wanted to bubble this review up. Everything should be good to go here. I've updated the settings.rst since there was a big refactor

justinmayer commented:
Many thanks to @rjames86 for the contribution and to @avaris for the review.  (^_^)
